src/rust/lib_ccxr/src/util/log.rs

```diff
@@ -0,0 +1,592 @@
+///
+/// # Usage
+/// The arguments works the same as a [`println!`] macro.
+///
+/// # Examples
+/// ```no_run
+/// # use lib_ccxr::util::log::*;
+/// info!("Processing the header section");
+/// ```
+#[macro_export]
+macro_rules! info {
+    ($($args:expr),*) => {
+        $crate::util::log::logger().expect("Logger is not yet initialized")
+            .log_info(&format_args!($($args),*))
+    };
+}
+
+/// Log a debug message.
+///
+/// Used for logging debug messages throughout the program. A logger needs to be setup initially
+/// by [`set_logger`].
+///
+/// # Usage
+/// This macro requires an [`DebugMessageFlag`] which represents the type of debug message. It is
+/// used for filtering the messages. This is provided using a key called `msg_type` which comes
+/// before the `;`. After `;`, the arguments works the same as a [`println!`] macro.
+///
+/// # Examples
+/// ```no_run
+/// # use lib_ccxr::util::log::*;
+/// # let byte1 = 23u8;
+/// # let byte2 = 45u8;
+/// debug!(
+///     msg_type = DebugMessageFlag::DECODER_708;
+///     "Packet Start with contents {} {}", byte1, byte2
+/// );
+/// ```
+#[macro_export]
+macro_rules! debug {
+    (msg_type = $msg_flag:expr; $($args:expr),*) => {
+        $crate::util::log::logger().expect("Logger is not yet initialized")
+            .log_debug($msg_flag, &format_args!($($args),*))
+    };
+}
+
+pub use debug;
+pub use error;
+pub use fatal;
+pub use info;
+
+/// Log a hex dump which is helpful for debugging purposes.
+///
+/// Setting `clear_high_bit` to true will ignore the highest bit whien displaying the
+/// characters. This makes visual CC inspection easier since the highest bit is usually used
+/// as a parity bit.
+///
+/// The byte numbers start from `0` by default. Use [`hex_dump_with_start_idx`] if a
+/// different starting index is required.
+pub fn hex_dump(message_type: DebugMessageFlag, data: &[u8], clear_high_bit: bool) {
+    logger()
+        .expect("Logger is not yet initialized")
+        .log_hex_dump(message_type, data, clear_high_bit, 0)
+}
+
+/// Log a hex dump which is helpful for debugging purposes.
+///
+/// Setting `clear_high_bit` to true will ignore the highest bit whien displaying the
+/// characters. This makes visual CC inspection easier since the highest bit is usually used
+/// as a parity bit.
+///
+/// The output will contain byte numbers which can be made to start from any number using
+/// `start_idx`. This is usually `0`.
+pub fn hex_dump_with_start_idx(
+    message_type: DebugMessageFlag,
+    data: &[u8],
+    clear_high_bit: bool,
+    start_idx: usize,
+) {
+    logger()
+        .expect("Logger is not yet initialized")
+        .log_hex_dump(message_type, data, clear_high_bit, start_idx)
+}
+
+/// Send a message to GUI.
+///
+/// Used for sending information related to XDS to the GUI.
+pub fn send_gui(message: GuiXdsMessage) {
+    logger()
+        .expect("Logger is not yet initialized")
+        .send_gui(message)
+}
+
+fn activity_header() {
```

Comment:
Now since there are two versions of `credits_shown` in the codebase, one in C and another in Rust. Imagine a function calling mprint in C and somewhere else in Rust, some function calls `info!`. Then activity header will be printed twice. To prevent this you will have to export `activity_header` to C just like I have done with other functions. And redirect all calls of C's `activity_header` to Rust's `activity_header`.

Reply:
Or you could remove the `CREDITS_SHOWN` variable completely and use the C's `credits_shown` variable in Rust since it is a static variable (as I did in #1556) although personally I don't like that approach. Probably wait for decision of @PunitLodha.

Reply:
As mentioned, it would be better if we just remove this credits shown check and also remove the `activity_header()` call from each `info!` log, and rather just print it once at the start of the program

Reply:
Did it